doc/user-manual/language/core-language.lagda.rst

```diff
@@ -8,21 +8,177 @@
+      | (x : a) → a             -- function space
+      | F                       -- data/record type
+      | c a                     -- data/record constructor
+      | s                       -- sort Seti, Setω+i
+
+
+Syntax overview
+---------------
+
+The syntax of an Agda program is defined in terms of three key components:
+
+* **Expressions** write function bodies and types.
+* **Declarations** declare types, data-types, postulates, records, functions etc.
+* **Pragmas** define program options.
+
+There are also three main levels of syntax, corresponding to different levels
+of interpretation:
+
+* **Concrete** is the high-level sugared syntax, it representing exactly what
+  the user wrote (Agda.Syntax.Concrete).
+* **Abstract**, before typechecking (Agda.Syntax.Abstract)
+* **Internal**, the full-intepreted core Agda terms, typechecked; roughly
+  corresponding to (Agda.Syntax.Internal).
+
+The process of translating an ``*.agda`` file into an executable has several
+stages:
+
+.. code-block:: text
+
+  *.agda file
+     ==[ parser (Lexer.x + Parser.y) ]==>
+  Concrete syntax
+     ==[ nicifier (Syntax.Concrete.Definitions) ]==>
+  'Nice' concrete syntax
+     ==[ scope checking (Syntax.Translation.ConcreteToAbstract) ]==>
+  Abstract syntax
+     ==[ type checking (TypeChecking.Rules.*) ]==>
+  Internal syntax
+     ==[ Agda.Compiler.ToTreeless ]==>
+  Treeless syntax
+     ==[ different backends (Compiler.MAlonzo.*, Compiler.JS.*, ...) ]==>
+  Source code
+     ==[ different compilers (GHC compiler, ...) ]==>
+  Executable
+
+The following sections describe these stages in more detail:
+
+Lexer
+-----
+
+.. _Alex: http://www.haskell.org/alex
+
+Lexical analysis (aka tokenization) is the process of converting a sequence of
+characters (the raw ``*.agda`` file) into a sequence of tokens (strings with a
+meaning).
+
+The lexer in Agda is generated by Alex_, and is an adaptation of GHC's lexer.
+The main lexing function ``lexer`` is called by the
+``Agda.Syntax.Parser.Parser`` to get the next token from the input.
+
+Parser
+------
+
+.. _Happy: http://www.haskell.org/happy
+
+The parser is the component that takes the output of the lexer and builds a
+data structure that we will call Concrete Syntax, while checking for correct
+syntax.
+
+The parser is generated by Happy_.
+
+Example: when a name is a sequence of parts, the lexer just sees it as a
+string, the parser does the translation in this step.
+
+
+Concrete Syntax
+---------------
+
+The concrete syntax is a raw representation of the program text without any
+desugaring at all.  This is what the parser produces. The idea is that if we
+figure out how to keep the concrete syntax around, it can be printed exactly
+as the user wrote it.
+
+Nice Concrete Syntax
+--------------------
+
+The translation from "Agda.Syntax.Concrete" to "Agda.Syntax.Abstract" involves
+scope analysis, figuring out infix operator precedences and tidying up
+definitions.
```

Comment:
This should probably go into the `Abstract Syntax` part.

The `Nice Concrete Syntax` is a slightly reorganised concrete syntax
that is easier to deal with internally. This is, among other things,
where new-style mutual blocks are detected, where definitions are
assembled from their isolated parts, where fixity information is
collected and attached to definitions, and where a lot of warnings
are emitted for valid but probably unintended things that are essentially
dead code such as empty `instance` blocks and misplaced pragmas.